plugins/modules/panos_readiness_checks.py

```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+#  Copyright 2023 Palo Alto Networks, Inc
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+DOCUMENTATION = """
+---
+module: panos_readiness_checks
+short_description: Runs readiness checks (boolean in nature) against a Firewall device.
+description:
+    - A wrapper around the PAN-OS Upgrade Assurance package.
+    - The module is meant to run readiness checks available in the package's L(CheckFirewall.run_readiness_checks()
+      method,https://pan.dev/panos/docs/panos-upgrade-assurance/api/check_firewall/#checkfirewallrun_readiness_checks).
+      Since it's just a wrapper, the way you would configure a check is exactly the same as if you would run the class directly.
+      Please refer to package's documentation for L(syntax,https://pan.dev/panos/docs/panos-upgrade-assurance/configuration-details/#readiness-checks)
+      and L(configuration dialect,https://pan.dev/panos/docs/panos-upgrade-assurance/dialect/).
+author: "Łukasz Pawlęga (@fosix)"
+version_added: '2.16.0'
+requirements:
+    - pan-python can be obtained from PyPI U(https://pypi.python.org/pypi/pan-python)
+    - pandevice can be obtained from PyPI U(https://pypi.python.org/pypi/pandevice)
+    - pan-os-upgrade-assurance can be obtained from PyPI U(https://pypi.org/project/panos-upgrade-assurance)
+notes:
+    - Only Firewalls are supported.
+    - Check mode is not supported.
+extends_documentation_fragment:
+    - paloaltonetworks.panos.fragments.transitional_provider
+    - paloaltonetworks.panos.fragments.vsys
+options:
+    checks:
+        description:
+            - A list of checks that should be run against a device. For the details on currently supported checks please refer to
+              L(package's documentation,https://pan.dev/panos/docs/panos-upgrade-assurance/configuration-details/#readiness-checks).
```

Comment:
true, but panos-upgrade-assurance is not for Ansible only, XSOAR can use it as well. Or just some pure Python script. Maybe more reasonable would be to keep it as part of the pan-os-python package?

Reply:
Ah, ok.  Separate it is, then!  So you'll want to add a new parameter to `./plugins/module_utils/panos.py` > `get_connection()` that allows you to do version checking against `panos_upgrade_assurance.version` the same as happens for both PAN-OS itself and pan-os-python / pandevice.  Please have the logic of this new check look like this:

```python
if self.min_panos_upgrade_assurance is not None:
    try:
        import panos_upgrade_assurance
    except ImportError:
        module.fail_json(msg='Missing required library "panos_upgrade_assurance".', syspath=sys.path)
    # This code assumes both panos_upgrade_assurance.version and self.min_panos_upgrade_assurance
    # are a tuple of 3 ints.  If panos_upgrade_assurance.version is a string, then you'll have
    # to turn it into a 3 element tuple of ints to do the comparison.
    if panos_upgrade_assurance.version < self.min_panos_upgrade_assurance:
        module.fail_json(msg=MIN_VERSION_ERROR.format(
            "panos_upgrade_assurance",
            _vstr(panos_upgrade_assurance.version),
            _vstr(self.min_panos_upgrade_assurance),
        ))
```

Reply:
@shinmog I am thinking we can use a super call in the FirewallProxy class to get the best of both worlds. Subclassing `FirewallProxy` makes life easy so what I'm thinking is something like (psuedocode here)

```python
firewall = helper.get_pandevice_parent().device
firewall_proxy = FirewallProxy(firewall)

# ... in assurance library ...

Class FirewallProxy(Firewall):
    def __init__(self, firewall_object: Optional[Firewall] = None, *args, **kwargs):
        """
        """
        if firewall_object:
            super(FirewallProxy, self).__init__(
                api_key=firewall_object.xapi.api_key,
                api_password=firewall_object.xapi.api_password,
                api_username=firewall_object.xapi.api_username,
                serial=firewall_object.xapi.serial,
                *args,
                **kwargs
            )
        else:
            super(FirewallProxy, self).__init__()
```

Can you see any reason this wouldn't work?

Reply:
finally `FirewallProxy` has both options, so it can take the `api-*` parameters or an existing device object. Changes will be introduced in upgrade assurance with [this pr](https://github.com/PaloAltoNetworks/pan-os-upgrade-assurance/pull/119).

This way we can leave the upgrade assurance as a *standalone* package and still use the `helper` class inside Ansible modules.

If someone has any objections, please leave a comment in the upgrade assurance PR.